compiler/rustc_mir_transform/src/shim.rs

```diff
@@ -1044,6 +1049,559 @@ fn build_fn_ptr_addr_shim<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId, self_ty: Ty<'t
+        /// ```
+        AsyncDropProjection,
+
+        /// Slice strategy should generate this:
+        ///
+        /// ```ignore (illustrative)
+        /// // unsizes array into a slice when needed
+        /// slice_async_destructor(to_drop)
+        /// ```
+        Slice { elem_ty: Ty<'tcx>, is_array: bool },
+
+        /// Chain strategy should generate something like this:
+        ///
+        /// ```ignore (illustrative)
+        /// chain(
+        ///     // Inserted only if surface AsyncDrop is implemented
+        ///     surface_async_drop_in_place(to_drop),
+        ///     chain(
+        ///         // We defer async drop of fields to defer creation of
+        ///         // mutable references not too early for the aliasing
+        ///         // model
+        ///         deferred_async_drop(addr_of_mut!((*to_drop).0)),
+        ///         chain(
+        ///             deferred_async_drop(addr_of_mut!((*to_drop).1)),
+        ///             chain(
+        ///                deferred_async_drop(addr_of_mut!((*to_drop).2)),
+        ///                ready_unit(), // for idempotency
+        ///             ),
+        ///         ),
+        ///     ),
+        /// )
+        /// ```
+        Chain { elem_tys: &'tcx ty::List<Ty<'tcx>>, has_surface_async_drop: bool },
+    }
+
+    let strategy = match *self_ty.kind() {
+        kind @ (ty::Array(elem_ty, _) | ty::Slice(elem_ty)) => {
+            GlueStrategy::Slice { is_array: matches!(kind, ty::Array(..)), elem_ty }
+        }
+
+        ty::Tuple(elem_tys) if !elem_tys.is_empty() => {
+            GlueStrategy::Chain { elem_tys, has_surface_async_drop: false }
+        }
+        ty::Adt(adt_def, args) if adt_def.is_struct() => {
+            debug_assert_eq!(adt_def.variants().len(), 1);
+            GlueStrategy::Chain {
+                elem_tys: tcx.mk_type_list_from_iter(
+                    adt_def.variant(VariantIdx::new(0)).fields.iter().map(|f| f.ty(tcx, args)),
+                ),
+                has_surface_async_drop: self_ty.is_async_drop(tcx, param_env),
+            }
+        }
+
+        ty::Foreign(_)
+        | ty::Adt(_, _)
+        | ty::Dynamic(_, _, _)
+        | ty::Closure(_, _)
+        | ty::CoroutineClosure(_, _)
+        | ty::Coroutine(_, _)
+        | ty::CoroutineWitness(_, _)
+        | ty::Alias(_, _)
+        | ty::Param(_)
+        | ty::Bound(_, _)
+        | ty::Placeholder(_)
+        | ty::Infer(_)
+        | ty::Error(_) => {
+            if self_ty.is_async_drop(tcx, param_env) {
+                GlueStrategy::AsyncDropProjection
+            } else {
+                GlueStrategy::Empty
+            }
+        }
+
+        ty::Bool
+        | ty::Char
+        | ty::Int(_)
+        | ty::Uint(_)
+        | ty::Float(_)
+        | ty::Str
+        | ty::RawPtr(_)
+        | ty::Ref(_, _, _)
+        | ty::FnDef(_, _)
+        | ty::FnPtr(_)
+        | ty::Never
+        | ty::Tuple(_) => GlueStrategy::Empty,
```

Comment:
shouldn't `Tuple` recurse into fields?

Reply:
This tuple is empty, case for non-empty tuple is about.

Reply:
I should probably reorder match arms to make this clear.